File: chapters/chp7_code/complex_functions.cpp
#include <iostream>
#include <complex>
#include <vector>
#include <cmath>
#include <fstream>
#include <map>
#include <string>

class ComplexFunction {
private:
    std::function<std::complex<double>(std::complex<double>)> func;
    
public:
    ComplexFunction(std::function<std::complex<double>(std::complex<double>)> f) : func(f) {}
    
    std::vector<std::complex<double>> find_zeros(double x_min, double x_max, double y_min, double y_max, 
                                                int resolution = 50, double tolerance = 1e-6) {
        std::vector<std::complex<double>> zeros;
        double dx = (x_max - x_min) / resolution;
        double dy = (y_max - y_min) / resolution;
        
        for (int i = 0; i <= resolution; ++i) {
            for (int j = 0; j <= resolution; ++j) {
                double x = x_min + i * dx;
                double y = y_min + j * dy;
                std::complex<double> z(x, y);
                
                try {
                    std::complex<double> w = func(z);
                    if (std::abs(w) < tolerance) {
                        zeros.push_back(z);
                    }
                } catch (...) {
                    // Skip points where function is undefined
                }
            }
        }
        
        return zeros;
    }
    
    std::vector<std::complex<double>> find_poles(double x_min, double x_max, double y_min, double y_max,
                                                int resolution = 50, double threshold = 1e6) {
        std::vector<std::complex<double>> poles;
        double dx = (x_max - x_min) / resolution;
        double dy = (y_max - y_min) / resolution;
        
        for (int i = 0; i <= resolution; ++i) {
            for (int j = 0; j <= resolution; ++j) {
                double x = x_min + i * dx;
                double y = y_min + j * dy;
                std::complex<double> z(x, y);
                
                try {
                    std::complex<double> w = func(z);
                    if (std::abs(w) > threshold) {
                        poles.push_back(z);
                    }
                } catch (...) {
                    poles.push_back(z);
                }
            }
        }
        
        return poles;
    }
    
    void save_analysis(double x_min, double x_max, double y_min, double y_max,
                      int resolution, const std::string& filename) {
        std::ofstream file(filename);
        file << "x,y,real_output,imag_output,magnitude,argument\n";
        
        double dx = (x_max - x_min) / resolution;
        double dy = (y_max - y_min) / resolution;
        
        for (int i = 0; i <= resolution; ++i) {
            for (int j = 0; j <= resolution; ++j) {
                double x = x_min + i * dx;
                double y = y_min + j * dy;
                std::complex<double> z(x, y);
                
                try {
                    std::complex<double> w = func(z);
                    double magnitude = std::abs(w);
                    double argument = std::arg(w);
                    
                    file << x << "," << y << "," << w.real() << "," << w.imag()
                         << "," << magnitude << "," << argument << "\n";
                } catch (...) {
                    file << x << "," << y << ",NaN,NaN,NaN,NaN\n";
                }
            }
        }
        
        file.close();
        std::cout << "Complex function analysis saved to " << filename << std::endl;
    }
};

// Define complex functions
std::map<std::string, std::function<std::complex<double>(std::complex<double>)>> create_function_library() {
    std::map<std::string, std::function<std::complex<double>(std::complex<double>)>> functions;
    
    // Polynomial functions
    functions["z"] = [](std::complex<double> z) { return z; };
    functions["z_squared"] = [](std::complex<double> z) { return z * z; };
    functions["z_cubed"] = [](std::complex<double> z) { return z * z * z; };
    functions["z_squared_plus_1"] = [](std::complex<double> z) { return z * z + 1.0; };
    
    // Rational functions
    functions["one_over_z"] = [](std::complex<double> z) { 
        if (std::abs(z) < 1e-10) throw std::runtime_error("Division by zero");
        return 1.0 / z; 
    };
    functions["one_over_z_squared_plus_1"] = [](std::complex<double> z) { return 1.0 / (z * z + 1.0); };
    functions["mobius"] = [](std::complex<double> z) { return (z - 1.0) / (z + 1.0); };
    
    // Transcendental functions
    functions["exp_z"] = [](std::complex<double> z) { return std::exp(z); };
    functions["sin_z"] = [](std::complex<double> z) { return std::sin(z); };
    functions["cos_z"] = [](std::complex<double> z) { return std::cos(z); };
    functions["log_z"] = [](std::complex<double> z) { 
        if (std::abs(z) < 1e-10) throw std::runtime_error("Log of zero");
        return std::log(z); 
    };
    
    // Special functions
    functions["sqrt_z"] = [](std::complex<double> z) { return std::sqrt(z); };
    functions["exp_one_over_z"] = [](std::complex<double> z) { 
        if (std::abs(z) < 1e-10) throw std::runtime_error("Essential singularity");
        return std::exp(1.0 / z); 
    };
    functions["z_times_exp_z"] = [](std::complex<double> z) { return z * std::exp(z); };
    
    return functions;
}

int main() {
    std::cout << "Complex Functions Analysis (C++)" << std::endl;
    std::cout << "================================" << std::endl;
    
    auto functions = create_function_library();
    
    // Analyze selected functions
    std::vector<std::string> selected = {"z_squared", "one_over_z", "exp_z", "sin_z", "mobius", "exp_one_over_z"};
    
    for (const auto& func_name : selected) {
        if (functions.find(func_name) != functions.end()) {
            std::cout << "Analyzing " << func_name << "..." << std::endl;
            
            ComplexFunction cf(functions[func_name]);
            
            // Choose appropriate domain
            double x_min = -3, x_max = 3, y_min = -3, y_max = 3;
            if (func_name == "exp_z") {
                y_min = -6; y_max = 6;
            } else if (func_name == "exp_one_over_z") {
                x_min = -1; x_max = 1; y_min = -1; y_max = 1;
            }
            
            // Save analysis
            std::string filename = "complex_function_" + func_name + "_cpp.csv";
            cf.save_analysis(x_min, x_max, y_min, y_max, 100, filename);
            
            // Find zeros and poles
            auto zeros = cf.find_zeros(x_min, x_max, y_min, y_max, 30);
            auto poles = cf.find_poles(x_min, x_max, y_min, y_max, 30);
            
            if (!zeros.empty()) {
                std::cout << "  Zeros found: " << zeros.size() << std::endl;
                for (size_t i = 0; i < std::min(size_t(5), zeros.size()); ++i) {
                    std::cout << "    " << zeros[i] << std::endl;
                }
            }
            
            if (!poles.empty()) {
                std::cout << "  Poles found: " << poles.size() << std::endl;
                for (size_t i = 0; i < std::min(size_t(5), poles.size()); ++i) {
                    std::cout << "    " << poles[i] << std::endl;
                }
            }
        }
    }
    
    std::cout << "\nFunction properties summary:" << std::endl;
    std::cout << "• Polynomial functions: Entire (analytic everywhere)" << std::endl;
    std::cout << "• Rational functions: Meromorphic (poles at zeros of denominator)" << std::endl;
    std::cout << "• Exponential functions: Entire, periodic in imaginary direction" << std::endl;
    std::cout << "• Trigonometric functions: Meromorphic, periodic" << std::endl;
    std::cout << "• Essential singularities: exp(1/z) has dense behavior near z=0" << std::endl;
    
    std::cout << "\nGenerated analysis files for " << selected.size() << " complex functions." << std::endl;
    std::cout << "Files can be visualized using domain coloring techniques." << std::endl;
    
    return 0;
}

File: chapters/chp7_code/complex_functions.py
#!/usr/bin/env python3

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import hsv_to_rgb
import warnings
warnings.filterwarnings('ignore')

class ComplexFunction:
    def __init__(self, function):
        self.function = function
    
    def evaluate_grid(self, x_range, y_range, resolution=200):
        x = np.linspace(*x_range, resolution)
        y = np.linspace(*y_range, resolution)
        X, Y = np.meshgrid(x, y)
        Z = X + 1j * Y
        
        try:
            W = self.function(Z)
            return X, Y, W
        except:
            return X, Y, np.full_like(Z, np.nan)
    
    def find_zeros_and_poles(self, x_range, y_range, resolution=50):
        X, Y, W = self.evaluate_grid(x_range, y_range, resolution)
        
        magnitude = np.abs(W)
        zeros = []
        poles = []
        
        for i in range(resolution):
            for j in range(resolution):
                if magnitude[i, j] < 1e-6:
                    zeros.append(complex(X[i, j], Y[i, j]))
                elif magnitude[i, j] > 1e6:
                    poles.append(complex(X[i, j], Y[i, j]))
        
        return zeros, poles
    
    def save_analysis_data(self, x_range, y_range, filename, resolution=200):
        X, Y, W = self.evaluate_grid(x_range, y_range, resolution)
        
        with open(filename, 'w') as f:
            f.write("x,y,real_output,imag_output,magnitude,argument,hue,brightness\n")
            
            for i in range(resolution):
                for j in range(resolution):
                    x, y = X[i, j], Y[i, j]
                    w = W[i, j]
                    
                    if np.isfinite(w):
                        magnitude = abs(w)
                        argument = np.angle(w)
                        hue = (argument + np.pi) / (2 * np.pi)
                        brightness = np.log(1 + magnitude)
                        
                        f.write(f"{x},{y},{w.real},{w.imag},{magnitude},{argument},{hue},{brightness}\n")
        
        print(f"Complex function analysis saved to {filename}")

# Define various complex functions
def polynomial_functions():
    return {
        'z': lambda z: z,
        'z²': lambda z: z**2,
        'z³': lambda z: z**3,
        'z⁴': lambda z: z**4,
        'z² + 1': lambda z: z**2 + 1,
        'z² - 1': lambda z: z**2 - 1,
        'z³ - z': lambda z: z**3 - z
    }

def rational_functions():
    return {
        '1/z': lambda z: np.where(z != 0, 1/z, np.inf),
        '1/(z²+1)': lambda z: 1/(z**2 + 1),
        '(z-1)/(z+1)': lambda z: (z-1)/(z+1),
        'z/(z²+1)': lambda z: z/(z**2 + 1),
        '(z²-1)/(z²+1)': lambda z: (z**2-1)/(z**2+1)
    }

def transcendental_functions():
    return {
        'e^z': lambda z: np.exp(z),
        'sin(z)': lambda z: np.sin(z),
        'cos(z)': lambda z: np.cos(z),
        'log(z)': lambda z: np.log(z + 1e-10),
        'sinh(z)': lambda z: np.sinh(z),
        'cosh(z)': lambda z: np.cosh(z),
        'tan(z)': lambda z: np.tan(z)
    }

def special_functions():
    return {
        'z^(1/2)': lambda z: np.sqrt(z + 1e-10),
        'z^(1/3)': lambda z: np.power(z + 1e-10, 1/3),
        'e^(1/z)': lambda z: np.exp(np.where(z != 0, 1/z, np.inf)),
        'sin(1/z)': lambda z: np.sin(np.where(z != 0, 1/z, np.inf)),
        'z*e^z': lambda z: z * np.exp(z),
        'z*sin(z)': lambda z: z * np.sin(z)
    }

if __name__ == "__main__":
    print("Complex Functions Analysis (Python version)")
    print("===========================================")
    
    # Analyze different categories of functions
    all_functions = {
        **polynomial_functions(),
        **rational_functions(), 
        **transcendental_functions(),
        **special_functions()
    }
    
    # Generate analysis for selected functions
    selected_functions = ['z²', '1/z', 'e^z', 'sin(z)', '(z-1)/(z+1)', 'e^(1/z)']
    
    for func_name in selected_functions:
        if func_name in all_functions:
            print(f"Analyzing {func_name}...")
            cf = ComplexFunction(all_functions[func_name])
            
            # Choose appropriate domain for each function
            if func_name in ['e^z']:
                x_range, y_range = (-2, 2), (-6, 6)
            elif func_name in ['sin(z)', 'cos(z)', 'tan(z)']:
                x_range, y_range = (-4, 4), (-4, 4)
            elif func_name in ['e^(1/z)', 'sin(1/z)']:
                x_range, y_range = (-1, 1), (-1, 1)
            else:
                x_range, y_range = (-3, 3), (-3, 3)
            
            # Save analysis
            safe_name = func_name.replace('/', '_div_').replace('^', '_pow_').replace('(', '').replace(')', '')
            filename = f"complex_function_{safe_name}_python.csv"
            cf.save_analysis_data(x_range, y_range, filename, resolution=150)
            
            # Find critical points
            zeros, poles = cf.find_zeros_and_poles(x_range, y_range, resolution=30)
            
            if zeros:
                print(f"  Zeros found: {len(zeros)}")
                for i, zero in enumerate(zeros[:5]):  # Show first 5
                    print(f"    {zero:.3f}")
            
            if poles:
                print(f"  Poles found: {len(poles)}")
                for i, pole in enumerate(poles[:5]):  # Show first 5
                    print(f"    {pole:.3f}")
    
    print("\nFunction properties summary:")
    print("• Polynomial functions: Entire (analytic everywhere)")
    print("• Rational functions: Meromorphic (poles at zeros of denominator)")
    print("• Exponential functions: Entire, periodic in imaginary direction")
    print("• Trigonometric functions: Meromorphic, periodic")
    print("• Logarithmic functions: Multi-valued, branch cuts")
    print("• Power functions: Multi-valued for non-integer powers")
    
    print(f"\nGenerated analysis files for {len(selected_functions)} complex functions.")
    print("Files can be visualized using domain coloring techniques.")